api/pkg/s3/objectcopy.go

```diff
@@ -15,15 +15,204 @@
+		} else {
+			WriteErrorResponse(response, request, ErrInvalidRequestBody)
+			return
+		}
+	}
+
+	log.Infoln("sourceBucketName:", sourceBucketName, " sourceObjectName:", sourceObjectName, " sourceVersion:", sourceVersion)
+
+	ctx := common.InitCtxWithAuthInfo(request)
+	sourceObject, err := s.getObjectMeta(ctx, sourceBucketName, sourceObjectName)
+	if err != nil {
+		log.Errorln("unable to fetch object info. err:", err)
+		WriteErrorResponse(response, request, err)
+		return
+	}
+
+	// Verify before x-amz-copy-source preconditions before continuing with CopyObject.
+	if err = checkObjectPreconditions(response.ResponseWriter, request.Request, sourceObject); err != nil {
+		WriteErrorResponse(response, request, err)
+		return
+	}
+
+	//TODO: In a versioning-enabled bucket, you cannot change the storage class of a specific version of an object. When you copy it, Amazon S3 gives it a new version ID.
+	storageClassFromHeader, err := getStorageClassFromHeader(request)
+	if err != nil {
+		WriteErrorResponse(response, request, err)
+		return
+	}
+	if storageClassFromHeader == types.ObjectStorageClassGlacier || storageClassFromHeader == types.ObjectStorageClassDeepArchive {
+		WriteErrorResponse(response, request, ErrInvalidCopySourceStorageClass)
+		return
+	}
+
+	// if source == dest and X-Amz-Metadata-Directive == REPLACE, only update the meta;
+	if isOnlyUpdateMetadata {
+		log.Infoln("only update metadata.")
+		targetObject := sourceObject
+
+		//update custom attrs from headers
+		newMetadata := extractMetadataFromHeader(request)
+		if c, ok := newMetadata["Content-Type"]; ok {
+			targetObject.ContentType = c
+		} else {
+			targetObject.ContentType = sourceObject.ContentType
+		}
+		targetObject.CustomAttributes = newMetadata
+		targetObject.StorageClass = storageClassFromHeader.ToString()
+
+		result, err := s.s3Client.UpdateObjectMeta(ctx, targetObject)
+		if err != nil {
+			log.Errorf("unable to update object meta for %v", targetObject.ObjectId)
+			WriteErrorResponse(response, request, err)
+			return
+		}
+		copyObjRes := GenerateCopyObjectResponse(result.Md5, time.Unix(result.LastModified, 0))
+		encodedSuccessResponse := EncodeResponse(copyObjRes)
+		// write headers
+		if result.Md5 != "" {
+			response.ResponseWriter.Header()["ETag"] = []string{"\"" + result.Md5 + "\""}
+		}
+		if sourceVersion != "" {
+			response.AddHeader("x-amz-copy-source-version-id", sourceVersion)
+		}
+		if result.VersionId != "" {
+			response.AddHeader("x-amz-version-id", result.VersionId)
+		}
+
+		log.Info("COPY object successfully.")
+		// write success response.
+		WriteSuccessResponse(response, encodedSuccessResponse)
+		return
+	}
+
+	/// maximum Upload size for object in a single CopyObject operation.
+	if isMaxObjectSize(sourceObject.Size) {
+		WriteErrorResponseWithResource(response, request, ErrEntityTooLarge, copySource)
+		return
+	}
+
+	// Note that sourceObject and targetObject are pointers
+	targetObject := &pb.Object{}
```

Comment:
It seems targetObject is not used after it is defined.

Reply:
It seems that you are right,  I will delete it in the next commit, thanks